### ISSUE TITLE
Under 2.2.0.0, windows file paths (~260 chars) fail to work properly via container mount

### ISSUE THREAD

**Issue by ephesused** [docker/for-win#5572]
- [ ] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior

Long file paths (over 260 characters) should be recognized and work properly.

### Actual behavior

Long file paths are not working properly in the container.

### Information

  - Windows Version: Windows 10 Enterprise, Version 1709, OS Build 16299.1565
  - Docker Desktop Version: 2.2.0.0 (42247)
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

### Steps to reproduce the behavior

```
c:\temp\testaaaaaa20\testaaa30\testaaa40\testaaa50\testaaa60\testaaa70\testaaa80\testaaa90\testaa100\testaa110\testaa120\testaa130\testaa140\testaa150\testaa160\testaa170\testaa180\testaa190\testaa200\testaa210\testaa220\testaa230\testaa240\testaa250>dir
 Volume in drive C is Windows
 Volume Serial Number is E469-F0BF

 Directory of c:\temp\testaaaaaa20\testaaa30\testaaa40\testaaa50\testaaa60\testaaa70\testaaa80\testaaa90\testaa100\testaa110\testaa120\testaa130\testaa140\testaa150\testaa160\testaa170\testaa180\testaa190\testaa200\testaa210\testaa220\testaa230\testaa240\testaa250

01/23/2020  10:43 AM    <DIR>          .
01/23/2020  10:43 AM    <DIR>          ..
01/23/2020  10:43 AM                 0 123
01/23/2020  10:43 AM                 0 123456
01/23/2020  10:43 AM                 0 1234567
01/23/2020  10:43 AM                 0 12345678
01/23/2020  10:42 AM                 0 123456789
01/23/2020  10:42 AM                 0 123456789012345
               6 File(s)              0 bytes
               2 Dir(s)  269,699,796,992 bytes free

c:\temp\testaaaaaa20\testaaa30\testaaa40\testaaa50\testaaa60\testaaa70\testaaa80\testaaa90\testaa100\testaa110\testaa120\testaa130\testaa140\testaa150\testaa160\testaa170\testaa180\testaa190\testaa200\testaa210\testaa220\testaa230\testaa240\testaa250>docker run --rm -it -v %CD%:/testing centos:7.6.1810
[root@5c8c198abcb5 /]# cd /testing
[root@5c8c198abcb5 testing]# ls -l
ls: cannot access 123456789: No such file or directory
ls: cannot access 123456789012345: No such file or directory
total 0
-rwxr-xr-x 1 root root 0 Jan 23 15:43 123
-rwxr-xr-x 1 root root 0 Jan 23 15:43 123456
-rwxr-xr-x 1 root root 0 Jan 23 15:43 1234567
-rwxr-xr-x 1 root root 0 Jan 23 15:43 12345678
?????????? ? ?    ?    ?            ? 123456789
?????????? ? ?    ?    ?            ? 123456789012345
[root@5c8c198abcb5 testing]# ls 12345678 123456789
ls: cannot access 123456789: No such file or directory
12345678
[root@5c8c198abcb5 testing]#
```

**Comment by ephesused** [comment]
For comparison, under 2.1.0.5 (40693):

```
c:\temp\testaaaaaa20\testaaa30\testaaa40\testaaa50\testaaa60\testaaa70\testaaa80\testaaa90\testaa100\testaa110\testaa120\testaa130\testaa140\testaa150\testaa160\testaa170\testaa180\testaa190\testaa200\testaa210\testaa220\testaa230\testaa240\testaa250>docker run --rm -it -v %CD%:/testing centos:7.6.1810
Unable to find image 'centos:7.6.1810' locally
7.6.1810: Pulling from library/centos
ac9208207ada: Pull complete
Digest: sha256:62d9e1c2daa91166139b51577fe4f4f6b4cc41a3a2c7fc36bd895e2a17a3e4e6
Status: Downloaded newer image for centos:7.6.1810
[root@8963f99c7927 /]# cd /testing
[root@8963f99c7927 testing]# ls -l
total 0
-rwxr-xr-x 1 root root 0 Jan 23 15:43 123
-rwxr-xr-x 1 root root 0 Jan 23 15:43 123456
-rwxr-xr-x 1 root root 0 Jan 23 15:43 1234567
-rwxr-xr-x 1 root root 0 Jan 23 15:43 12345678
-rwxr-xr-x 1 root root 0 Jan 23 15:42 123456789
-rwxr-xr-x 1 root root 0 Jan 23 15:42 123456789012345
[root@8963f99c7927 testing]# ls 12345678 123456789
12345678  123456789
[root@8963f99c7927 testing]#
```

**Comment by sakilo1990** [comment]
I am having the same issue, had to revert back to a previous version

**Comment by mat007** [comment]
Thanks for reporting this issue, we are working on a fix.

**Comment by G-Rath** [comment]
@mat007 don't mean to pressure you, but is there an approximate ETA on when this might be fixed?

I'm getting yarn install issues which I believe are the result of this:

> error An unexpected error occurred: "https://registry.yarnpkg.com/@babel/helper-builder-binary-assignment-operator-visitor/-/helper-builder-binary-assignment-operator-visitor-7.1.0.tgz: EIO: i/o error, open '/var/www/.cache/yarn/v6/npm-@babel-helper-builder-binary-assignment-operator-visitor-7.1.0-6b69628dfe4087798e0c4ed98e3d4a6b2fbd2f5f/node_modules/@babel/helper-builder-binary-assignment-operator-visitor/.yarn-tarball.tgz'

Attempting to create that directory results in the behaviour described in the above.

Just a rough ballpark on if we're talking weeks or months would be great :)

**Comment by G-Rath** [comment]
Just to add some more info: I have gotten it at path lengths above 200 but less than 260 (but the same packages, which is interesting).

Can confirm that downgrading to previous version (2.1.0.0) fixed it right up for me.

**Comment by mat007** [comment]
> @mat007 don't mean to pressure you, but is there an approximate ETA on when this might be fixed?

We’re looking at another release in about a week or so.

**Comment by stephen-turner** [comment]
We have a new private build that (we believe) fixes this bug. https://download-stage.docker.com/win/stable/42579/Docker%20Desktop%20Installer.exe. Please try it and let us know how you get on.

**Comment by G-Rath** [comment]
@stephen-turner I'm trying it out now - so far the results are promising.

I'm going to retry with a `docker system prune --all` just to be sure, but I think you've got :)

**Comment by G-Rath** [comment]
@stephen-turner so the file name length problem seems to have been successfully resolved.

However, I have found another bug with the new file system setup that I've confirmed doesn't happen on 2.1:

![image](https://user-images.githubusercontent.com/3151613/73577708-84d3eb80-44e2-11ea-9376-766f035fc75e.png)

Effectively, if I remove folders on the host side, apps like `npm` & `composer` still think that they're there and fail, but I can't `ls`, `stat` etc those files. 

I've confirmed that problem remains until the Docker daemon is restarted (i.e `docker-compose down -v` & `docker-compose up` doesn't resolve it).

Let me know if you'd like me to make a new issue, or if this is known :)

https://gist.github.com/G-Rath/c351b12b456ad74a13eecd1bf23e383f
